enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,964 @@
+        "name": "etcdoperator.v0.9.0",
+        "package": "etcd",
+        "image": "quay.io/operatorhubio/etcd:v0.9.0",
+        "version": "0.9.0",
+        "properties":[
+            {
+                "name": "olm.package",
+                "values": {
+                    "packageName": "etcd",
+                    "version": "0.9.0"
+                }
+            },
+            {
+                "name":"olm.gvk",
+                "values":{
+                    "group": "etcd.database.coreos.com", 
+                    "kind": "EtcdBackup", 
+                    "version": "v1beta2"
+                }
+            }
+        ],
+        "channels": [
+            {
+                "name": "singlenamespace-alpha",
+                "upgradeInfo": {}
+            },
+        ],
+        "relatedImages" : [
+            {
+                "name": "etcdv0.9.0",
+                "image": "quay.io/coreos/etcd-operator@sha256:db563baa8194fcfe39d1df744ed70024b0f1f9e9b55b5923c2f3a413c44dc6b8"
+            }
+        ]
+    }
+]
+```
+
+### Using the package json representation
+
+Once every package inside an index is represented using a json/yaml config file, the config files can be leveraged to inspect/update/validate the packages inside the index without needing to rebuild the index from scratch.  
+#### Inspecting the packages inside an index
+
+A new sub-command `inspect` will be introduced under `opm index`.
+
+When `opm index inspect` is summoned, the json representations of the packages inside the index will be unfurled in a folder with the same name as the index. 
+
+```bash 
+$ opm index inspect --index=docker.io/my-namespace/community-operators --output=json
+$ cd community-operators
+$ tree
+.
+├── etcd.json
+├── amqstreams.json
+├── image_sha256.json
+```
+#### Editing a package inside an index
+
+A new sub-command `update` will be introduced under `opm index`, which will take a folder containing json representations of packages as it's input along with an existing index image tag, and will push any edit made to the config files up to the index container image. 
+
+```bash
+$ opm index inspect --index=quay.io/some-namespace/my-index 
+$ // edit a my-index/package-config.json 
+$ opm index update my-index --tag=docker.io/some-namespace/my-index:latest
+```
+
+When an index is "inspected" for it's config files, a `image_sha256.json` will also be created that will store the sha of the index image being inspected. 
+```bash 
+$ cat community-operators/image_sha256.json
+{
+    "sha256":"9e0c9c2c8e973c7e17439509d1dc1512afa3a8ed7c0b169016299a357bbc5b10"
+}
+```
+When an update is attempted on an index image, the `update` command will verify that the sha information locally matches with the sha for the remote index image before overriding the content of the container image with the latest changes. This will prevent overriding of existing information that may not have been pulled locally due to asynchronous update of the config files. 
```

Comment:
I'm torn on this idea.

I could see this being a useful feature to help prevent accidental overwrites.

On the other hand, indexes are "just" images, and "just" json files. There's nothing unique about preventing overwrites to an index image vs. any other type of image. I'm not convinced this is a task that opm needs to be concerned with (since there are plenty of other ways to solve it that would be specific to a particular workflow / consumer).

(Also, this feature doesn't do a "real" lock, and performing this check ahead of time doesn't actually guarantee that another actor hasn't swapped the tag in between when you've checked that value and when you've pushed)

Reply:
It just feels like a gap not to provide a mechanism to ensure data integrity of an artifact that can be modified concurrently by distributed players . 

> Also, this feature doesn't do a "real" lock, and performing this check ahead of time doesn't actually guarantee that another actor hasn't swapped the tag in between when you've checked that value and when you've pushed

That is true, and I don't think it was aiming to serve as a real lock but serve as an accidental override prevention mechanism. 

Having said that, if we think it'll do more harm than good, I can remove it for now and maybe think about it later if we get feedback of accidental overrides being a nuisance.

Reply:
I'd be in favor of removing it. I think it gives the illusion of safety without actual safety. No reason to overcommit here.

Reply:
i'm +1 on removal mostly because it will make the pipeline implementation easier, i think.